crypto/evp/evp_exchange.c

```diff
@@ -0,0 +1,354 @@
+}
+
+static void *evp_exchange_from_dispatch(const OSSL_DISPATCH *fns,
+                                        OSSL_PROVIDER *prov)
+{
+    EVP_EXCHANGE *exchange = NULL;
+    int fncnt = 0;
+
+    if ((exchange = evp_exchange_new(prov)) == NULL)
+        return NULL;
+
+    for (; fns->function_id != 0; fns++) {
+        switch (fns->function_id) {
+        case OSSL_FUNC_EXCHANGE_NEWCTX:
+            if (exchange->newctx != NULL)
+                break;
+            exchange->newctx = OSSL_get_OP_exchange_newctx(fns);
+            fncnt++;
+            break;
+        case OSSL_FUNC_EXCHANGE_INIT:
+            if (exchange->init != NULL)
+                break;
+            exchange->init = OSSL_get_OP_exchange_init(fns);
+            fncnt++;
+            break;
+        case OSSL_FUNC_EXCHANGE_SET_PEER:
+            if (exchange->set_peer != NULL)
+                break;
+            exchange->set_peer = OSSL_get_OP_exchange_set_peer(fns);
+            break;
+        case OSSL_FUNC_EXCHANGE_DERIVE:
+            if (exchange->derive != NULL)
+                break;
+            exchange->derive = OSSL_get_OP_exchange_derive(fns);
+            fncnt++;
+            break;
+        case OSSL_FUNC_EXCHANGE_FREECTX:
+            if (exchange->freectx != NULL)
+                break;
+            exchange->freectx = OSSL_get_OP_exchange_freectx(fns);
+            fncnt++;
+            break;
+        case OSSL_FUNC_EXCHANGE_DUPCTX:
+            if (exchange->dupctx != NULL)
+                break;
+            exchange->dupctx = OSSL_get_OP_exchange_dupctx(fns);
+            break;
+        case OSSL_FUNC_EXCHANGE_SET_PARAMS:
+            if (exchange->set_params != NULL)
+                break;
+            exchange->set_params = OSSL_get_OP_exchange_set_params(fns);
+            break;
+        }
+    }
+    if (fncnt != 4) {
+        /*
+         * In order to be a consistent set of functions we must have at least
+         * a complete set of "exchange" functions: init, derive, newctx,
+         * and freectx. The dupctx, set_peer and set_params functions are
+         * optional.
+         */
+        EVP_EXCHANGE_free(exchange);
+        EVPerr(EVP_F_EVP_EXCHANGE_FROM_DISPATCH, EVP_R_INVALID_PROVIDER_FUNCTIONS);
+        return NULL;
+    }
+
+    return exchange;
+}
+
+void EVP_EXCHANGE_free(EVP_EXCHANGE *exchange)
+{
+    if (exchange != NULL) {
+        int i;
+
+        CRYPTO_DOWN_REF(&exchange->refcnt, &i, exchange->lock);
+        if (i > 0)
+            return;
+        ossl_provider_free(exchange->prov);
+        CRYPTO_THREAD_lock_free(exchange->lock);
+        OPENSSL_free(exchange);
+    }
+}
+
+int EVP_EXCHANGE_up_ref(EVP_EXCHANGE *exchange)
+{
+    int ref = 0;
+
+    CRYPTO_UP_REF(&exchange->refcnt, &ref, exchange->lock);
+    return 1;
+}
+
+EVP_EXCHANGE *EVP_EXCHANGE_fetch(OPENSSL_CTX *ctx, const char *algorithm,
+                                 const char *properties)
+{
+    EVP_EXCHANGE *exchange =
```

Comment:
Is local variable needed here..
either just return, or get rid of extra blankline

Reply:
Removed the local var